src/librustc_typeck/check/_match.rs

```diff
@@ -89,30 +89,46 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
             let lhs_ty = fcx.expr_ty(&**begin);
             let rhs_ty = fcx.expr_ty(&**end);
 
-            let lhs_eq_rhs =
+            let compat = |t| ty::type_is_numeric(t) || ty::type_is_char(t);
+            let lhs_compat = compat(lhs_ty);
+            let rhs_compat = compat(rhs_ty);
+
+            if !lhs_compat || !rhs_compat {
+
+                let span = if !lhs_compat && !rhs_compat {
+                    pat.span
+                } else if !lhs_compat {
+                    begin.span
+                } else {
+                    end.span
+                };
+
+                span_err!(tcx.sess, span, E0029,
+                          "only char and numeric types are allowed in range");
+                span_note!(tcx.sess, span, "Start type: {}\nEnd type: {}",
+                    fcx.infcx().ty_to_string(lhs_ty),
+                    fcx.infcx().ty_to_string(rhs_ty)
+                );
+                return;
+            }
+
+            let types_match =
                 require_same_types(
-                    tcx, Some(fcx.infcx()), false, pat.span, lhs_ty, rhs_ty,
+                    tcx, Some(fcx.infcx()), false, end.span, rhs_ty, lhs_ty,
                     || "mismatched types in range".to_string());
 
-            let numeric_or_char =
-                lhs_eq_rhs && (ty::type_is_numeric(lhs_ty) || ty::type_is_char(lhs_ty));
+            if !types_match {
+                return;
+            }
 
-            if numeric_or_char {
-                match const_eval::compare_lit_exprs(tcx, &**begin, &**end, Some(lhs_ty)) {
-                    Some(Ordering::Less) |
-                    Some(Ordering::Equal) => {}
-                    Some(Ordering::Greater) => {
-                        span_err!(tcx.sess, begin.span, E0030,
-                            "lower range bound must be less than upper");
-                    }
-                    None => {
-                        span_err!(tcx.sess, begin.span, E0031,
-                            "mismatched types in range");
-                    }
+            match const_eval::compare_lit_exprs(tcx, &**begin, &**end, Some(lhs_ty)) {
+                Some(Ordering::Less) |
+                Some(Ordering::Equal) => {}
+                Some(Ordering::Greater) => {
+                    span_err!(tcx.sess, begin.span, E0030,
+                        "lower range bound must be less than or equal to upper");
                 }
-            } else {
-                span_err!(tcx.sess, begin.span, E0029,
-                          "only char and numeric types are allowed in range");
+                None => unreachable!()
             }
 
             fcx.write_ty(pat.id, lhs_ty);
```

Comment:
I think you could write this type in the error cases too, it should let you type check further.